### PR TITLE
feat: VIP for VLANs

### LINKED ISSUE #3601: Control plane shared VIP on a VLAN interface.
## Feature Request

I just ran into this on a lab cluster, where I'd like to have the shared vip on a vlan. Right now its just possible to put it on a "main" interface, though.

I'm not sure what we'd like the UX to be like though, since it doesn't make much sense to have more than one (?). Maybe move the control plane vip config outside the interface config(s) and keep it separate?

### PR DESCRIPTION
**Pull request by frezbo**
Support VIP for VLANS

Fixes #3601

---
This change is [Reviewable](https://reviewable.io/reviews/talos-systems/talos/4633)

### CODE REVIEW

**Comment by frezbo** [comment]
Tested on the local QEMU cluster

- `10.5.0.2` is the node IP
- `10.5.0.10` VIP assigned to `eth0`
- `10.5.0.20` VIP assigned to `eth0.100` VLAN

```bash
❯ talosctl -n 10.5.0.2 get addresses
NODE       NAMESPACE   TYPE            ID                                          VERSION   ADDRESS                        LINK
10.5.0.2   network     AddressStatus   cni0/10.244.0.1/24                          1         10.244.0.1/24                  cni0
10.5.0.2   network     AddressStatus   cni0/fe80::5404:9cff:fecf:4fc5/64           2         fe80::5404:9cff:fecf:4fc5/64   cni0
10.5.0.2   network     AddressStatus   eth0.100/10.5.0.20/32                       1         10.5.0.20/32                   eth0.100
10.5.0.2   network     AddressStatus   eth0.100/fe80::24bb:f9ff:fe2c:4e57/64       2         fe80::24bb:f9ff:fe2c:4e57/64   eth0.100
10.5.0.2   network     AddressStatus   eth0/10.5.0.10/32                           1         10.5.0.10/32                   eth0
10.5.0.2   network     AddressStatus   eth0/10.5.0.2/24                            1         10.5.0.2/24                    eth0
10.5.0.2   network     AddressStatus   eth0/fe80::24bb:f9ff:fe2c:4e57/64           2         fe80::24bb:f9ff:fe2c:4e57/64   eth0
10.5.0.2   network     AddressStatus   flannel.1/10.244.0.0/32                     1         10.244.0.0/32                  flannel.1
10.5.0.2   network     AddressStatus   flannel.1/fe80::fc30:cfff:fe55:790a/64      2         fe80::fc30:cfff:fe55:790a/64   flannel.1
10.5.0.2   network     AddressStatus   lo/127.0.0.1/8                              1         127.0.0.1/8                    lo
10.5.0.2   network     AddressStatus   lo/::1/128                                  1         ::1/128                        lo
10.5.0.2   network     AddressStatus   veth4eaa8cdf/fe80::3ce2:8cff:fe63:5538/64   2         fe80::3ce2:8cff:fe63:5538/64   veth4eaa8cdf
10.5.0.2   network     AddressStatus   veth957715e4/fe80::d82a:bcff:fe0c:3fdd/64   2         fe80::d82a:bcff:fe0c:3fdd/64   veth957715e4

❯ talosctl -n 10.5.0.2 get operatorspecs
NODE       NAMESPACE   TYPE           ID               VERSION
10.5.0.2   network     OperatorSpec   dhcp4/eth0       1
10.5.0.2   network     OperatorSpec   dhcp4/eth0.100   1
10.5.0.2   network     OperatorSpec   vip/eth0         1
10.5.0.2   network     OperatorSpec   vip/eth0.100     1

❯ ping -c 3 10.5.0.10
PING 10.5.0.10 (10.5.0.10) 56(84) bytes of data.
64 bytes from 10.5.0.10: icmp_seq=1 ttl=64 time=0.548 ms
64 bytes from 10.5.0.10: icmp_seq=2 ttl=64 time=0.520 ms
64 bytes from 10.5.0.10: icmp_seq=3 ttl=64 time=0.393 ms

--- 10.5.0.10 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2056ms
rtt min/avg/max/mdev = 0.393/0.487/0.548/0.067 ms

❯ ping -c 3 10.5.0.20
PING 10.5.0.20 (10.5.0.20) 56(84) bytes of data.
64 bytes from 10.5.0.20: icmp_seq=1 ttl=64 time=0.568 ms
64 bytes from 10.5.0.20: icmp_seq=2 ttl=64 time=0.407 ms
64 bytes from 10.5.0.20: icmp_seq=3 ttl=64 time=0.551 ms

--- 10.5.0.20 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2059ms
rtt min/avg/max/mdev = 0.407/0.508/0.568/0.072 ms
❯ 
```

File: internal/app/machined/pkg/controllers/network/operator_config.go [hunk 1]
```diff
@@ -308,3 +280,47 @@ func (ctrl *OperatorConfigController) apply(ctx context.Context, r controller.Ru
 
 	return ids, nil
 }
+
+func handleVIP(ctx context.Context, vlanConfig talosconfig.VIPConfig, deviceName string, logger *zap.Logger) (*network.OperatorSpecSpec, error) {
```
Comment (smira) [inline internal/app/machined/pkg/controllers/network/operator_config.go:284]:
```suggestion
func handleVIP(ctx context.Context, vlanConfig talosconfig.VIPConfig, deviceName string, logger *zap.Logger) (network.OperatorSpecSpec, error) {
```

Comment (smira) [reply internal/app/machined/pkg/controllers/network/operator_config.go:284]: That is more of a nit, but as we seem to be using the return value by value always, we can return by value as well. Go is pretty good at copying values, and it should anyways stay on the stack.

**Review by smira — COMMENTED** [review]
lgtm, with one small nit

we could also highlight new features on VLANs in `hack/release.toml`. it might be something short saying that now Talos supports MTU and Virtual IPs on VLAN interfaces

This way it will go to the release notes

File: internal/app/machined/pkg/controllers/network/operator_config.go (hunk 1 above)
Comment (frezbo) [reply internal/app/machined/pkg/controllers/network/operator_config.go]: fixed

**Review by smira — APPROVED** [review]
:+1:

**Pull request merged**